### ISSUE TITLE
Getting Room Name or Subject From Muc Hooks

### ISSUE THREAD

**Issue by tolgatanriverdi** [processone/ejabberd#1384]
I want to get the room's name or subject from one of the ejabberd hooks.
I've a method like in the below code and I want to get room's name or subject within that method.
Is something like that possible?

    muc_filter_message(Stanza, MUCState, RoomJID, FromJID, FromNick) ->
        PostUrl = gen_mod:get_module_opt(FromJID#jid.lserver, ?MODULE, post_url, fun(S) -> iolist_to_binary(S) end, list_to_binary("")),
        Token = gen_mod:get_module_opt(FromJID#jid.lserver, ?MODULE, auth_token, fun(S) -> iolist_to_binary(S) end, list_to_binary("")),
        Body = fxml:get_path_s(Stanza, [{elem, list_to_binary("body")}, cdata]),
    
        _LISTUSERS = lists:map(
            fun({_LJID, Info}) ->
                binary_to_list(Info#user.jid#jid.luser) ++ ".."
            end,
            dict:to_list(MUCState#state.users)
        ),
        ?DEBUG(" #########    GROUPCHAT _LISTUSERS = ~p~n  #######   ", [_LISTUSERS]),
    
        _AFILLIATIONS = lists:map(
            fun({{Uname, _Domain, _Res}, _Stuff}) ->
                binary_to_list(Uname) ++ ".."
            end,
            dict:to_list(MUCState#state.affiliations)
        ),
        ?DEBUG(" #########    GROUPCHAT _AFILLIATIONS = ~p~n  #######   ", [_AFILLIATIONS]),
    
        _OFFLINE = lists:subtract(_AFILLIATIONS, _LISTUSERS),
        ?DEBUG(" #########    GROUPCHAT _OFFLINE = ~p~n  #######   ", [_OFFLINE]),
    
        if
            Stanza /= "", length(_OFFLINE) > 0 ->
                Sep = "&",
                Post = [
                    "type=groupchat", Sep,
                    "to=", RoomJID#jid.luser, Sep,
                    "from=", FromJID#jid.luser, Sep,
                    "offline=", _OFFLINE, Sep,
                    "nick=", FromNick, Sep,
                    "body=", url_encode(binary_to_list(Body)), Sep,
                    "access_token=", Token
                ],
                ?INFO_MSG("Sending post request to ~s with body \"~s\"", [PostUrl, Post]),
                httpc:request(post, {binary_to_list(PostUrl), [], "application/x-www-form-urlencoded", list_to_binary(Post)},[],[]),
                Stanza;
            true ->
                Stanza
        end.

Thanks

**Comment by badlop** [comment]
Try something like this:
```
    Roomname = MUCState#state.room,
    Subject = MUCState#state.subject,
    Config = MUCState#state.config,
    Title = Config#config.title,
    ?INFO_MSG("Room Name: ~p, Title: ~p, Subject: ~p.", [Roomname, Title, Subject]),
```

**Comment by tolgatanriverdi** [comment]
Thanks, this worked

```
    Config = MUCState#state.config,
    Title = Config#config.title,
```

**Comment by tolgatanriverdi** [comment]
I've found a problem about this solution.
In a room we are making our tests there are 9 people in the member list and at the time that we're sending message,  7 of them are offline.
But when we send the message, mod_offline forwards only 6 of them , it acts 1 user as still online even though he is not.(He'is disconnected from ejabberd)
Do you know how can we solve that problem

You can find the logs in below. If you need any more info I can provide the details

`Type:groupchat Offline:905551112233..905326650263..905379514986..905558429874..16503312600..905326400417.. Room_Name:testroom`

**Comment by tolgatanriverdi** [comment]
any idea?

**Comment by badlop** [comment]
Check if the problem is reproducible: does it happen always for the same account? or does it happen for the last one/first one that left the room? Try with a stock ejabberd (just installed, without customizations). I mean, try to reduce the amount of variables as much as possible, so that you can slowly get an idea where the problem can be.

**Comment by tolgatanriverdi** [comment]
It looks like its always happening for only 1 user. That is mostly the person who became offline last

**Comment by tolgatanriverdi** [comment]
Is there any way to debug this problem and send the related logs to you. Because this became an urgent matter for us.
Can you help me a little bit on this issue?

**Comment by badlop** [comment]
I tried to reproduce the problem: installed ejabberd 16.12-beta1, Created two accounts, one logins and creates the room, the other logins and joins the room, subscribes to it, then leaves the room and logouts. When the room owner sends a message to the room, the message stanza is stored in the user's offline (I see it in the WebAdmin).

Can you try this? If it doesn't work in your current server, disable your custom code, or install 16.12-beta (or 16.09) and check if it works.

In summary: either you found a very strange bug that I can't reproduce, or your custom code is interfering with normal behavior and is provoking somehow that missing message.

**Comment by tolgatanriverdi** [comment]
This problem is not related with the other one (#1389) 
This is another problem.
This problem happens like this

1. The admin creates the room
2. Admin invites 4 people and add them to members list
3. Invited members join and subscribe to room
3. All users can send and receive message succesfully
4. 3 of the members became offline
5. 1 online user send message to group
6. 2 of the members receive push notification about the message(Because of mod_offline_post)
7. 1 user doesnt receive push notification because mod_offline_post doesnt see that user as offline.(This is always the last user who became offline)

**Comment by badlop** [comment]
What is mod_offline_post?

**Comment by tolgatanriverdi** [comment]
The module for forwarding messages and muc messages which is in the first message of this thread

**Comment by badlop** [comment]
Try disabling it, enabling mod_offline instead, and check if the messages are stored by mod_offline correctly. If not, then there may be a bug in ejabberd, otherwise there's a bug in that code.

**Comment by tolgatanriverdi** [comment]
I've disabled the mod_offline_post and still the same result. If a member de-members himself, the message related with that activity cannot be sent to offline users.
How can I fix that problem?

**Comment by badlop** [comment]
> I've disabled the mod_offline_post and still the same result.

And have you mod_offline enabled, with the default configuration?

> If a member de-members himself, the message related with that activity cannot be sent to offline users.

*Cannot* be sent? The message is caught by mod_offline, stored in the database, and sent later when the user logins.

> How can I fix that problem?

There is still one suggestion I made that you can try: install ejabberd 16.09 in another machine, make only small configuration changes (IP listener address, and enable mod_register), create a few accounts and make your test. Does it work correctly, or does it fail?

**Comment by tolgatanriverdi** [comment]
Ok I'll give it a try and tell you the results of it

**Comment by tolgatanriverdi** [comment]
I've installed 16.12 beta 1 version and the result is still the same:

User-1 of the group de-member himself with using
<iq type='set' to='sala1@conference.localhost' id='1232312312'>
  <query xmlns='http://jabber.org/protocol/muc#admin'>
    <item nick='user2' affiliation='none'/>
  </query>
</iq>
User-2 of the group never receives a message about the user-1 action

**Comment by badlop** [comment]
> I've installed 16.12 beta 1 version and the result is still the same:
> User-1 of the group de-member himself successfully 

How can you do that? If I install 16.12-beta1, create accounts, login to them, create room, member 1, subscribe 2, then demember 1 himself.. then ejabberd complains: 405 not-allowed.
